Documents/AnalyticsToolDesignDocument.md

```diff
@@ -0,0 +1,48 @@
+- **File system** - Provide enumerated details for files in the placeholder list or the modified path list
+
+## Background Information
+
+- **Modified paths** is the list of files which VFS4G has completely pulled from the remote and whose contents are on disk in their entirety. These files are handed to git via a hook when any git command requiring the list of changed files is called
+- **Placeholders** are files which have been opened by the local system with a read tag but without a write tag. They are objects tracked by VFS4G since git should not have to internally scan them since it is guaranteed that they have not been modified and accordingly do not have to be rehashed for differences. Their contents are on disk, but control over them has not been entirely ceded over to the system. They decrease performance because VFS4G has to track all of them individually, and this can cause certain `gvfs` commands to be slow
+- The **Git index** is gits internal list of files which have been modified (tracked or untracked) which is what is used for commit commands to keep score of everything going on inside of the enlistment. It is useful to this project by means of comparing it to the list of **hydrated files** to deduce if those files should truly be hydrated or if it may have occurred incidentally.
+- The **file system** must be carefully considered in the case of working with a mounted repository. VFS4G intercepts file I/O calls at the kernel level to inject its own behavior (the projected file system and if needed a network call to pull down the file which the system asked for in the correct format). Because of this, enumerating directories can cause additional data to be pulled down from the remote by hydrating extra files or further hydrating those that already exist on disk. Accordingly, care must be taken when interacting with the file system lest the tool designed to help users increase enlistment heath end up causing more problems.
+
+## Components
+
+- **API** - The server component of the client-server relationship. Responsible for the logic and computation of statistical data, as well as the assembly and sending of a **json statistics object** over the named pipe to the client. It sits within the `InProgressMount` process and interacts with the VFS4G internals
+- **Verb** - The `gvfs statistics` verb will be the component that handles the client side interaction with the API. When run, it will collect and store the statistics from the enlistment to either be displayed immediately or collected for trend purposes to then provide understanding in the future. This is also the point where data can be sent to telemetry for our internal analysis
+- **CLI Tool** - This tool will be built into the above verb and used to visualize and understand the data. It will simply output the analytics data to the command line and will also be used during the design period to test the API and verb

```

Comment:
I'm confused about the distinction between the **Verb** and the **CLI Tool**. A verb is specifically what executes when we run `gvfs <verbname>` in the command-line.

Reply:
I'm envisioning the verb having overall functionality to do other things, I'm not sure if something like the verb is actually used when running on a timer to collect data (perhaps that is a different type of system in place like you had mentioned some yesterday). I envision the CLI tool (looking back, that's an ambiguous name) being an option for the verb which attempts to present the information to the user in a somewhat beautified manner. Perhaps the two are really the same thing and the default action that occurs when you run `gvfs statistics` is something like you showed above or the mock up I created and will push shortly. Then those two are really just one and the same which makes sense

Reply:
If you are thinking instead about the object model, then it would be something like:

* `StatisticsVerb`
  - Lives in `GVFS/GVFS/CommandLine`
  - Handles all parameter handling and user-facing UI decisions
* `StatisticsWorker` (or something)
  - Lives in `GVFS/GVFS.Common/Statistics`
  - Runs multiple `StatisticsJob` instances, say `PlaceholderStatisticsJob` and `ModifiedFilesStatisticsJob`.

Then, you are talking about the difference between the command-line interface (verb) and the business logic (worker). The nice thing about the split here is that you can have the mount process also run the statistics on a regular basis and output the results a different way (telemetry, platform native notifications).

Reply:
Yeah splitting them up that way makes sense to me. I need to think some about how I'll refactor the document but I definitely think this fits more closely with the model I had in my head than the one I had on paper.